class_name Game extends Node2D

const player_definition : EntityDefinition = preload("res://assets/definitions/entities/actors/entity_definition_player.tres")

@onready var player: Entity
@onready var event_handler: EventHandler = $EventHandler
@onready var entities: Node2D = $Entities
@onready var map: Node2D = $Map


func _ready() -> void:
	player = Entity.new(Vector2i.ZERO, player_definition)
	var camera: Camera2D = $Camera2D
	remove_child(camera)
	player.add_child(camera)
	entities.add_child(player)
	map.generate(player)
	#var player_start_pos: Vector2i = Grid.world_to_grid(get_viewport_rect().size.floor() / 2)
	#player = Entity.new(player_start_pos, player_definition)
	#entities.add_child(player)
	#var npc := Entity.new(player_start_pos + Vector2i.RIGHT, player_definition)
	#npc.modulate = Color.ORANGE_RED
	#entities.add_child(npc)

func get_map_data() -> MapData:
	return map.map_data

func _physics_process(_delta: float) -> void:
	var action: Action = event_handler.get_action()
	if action:
		action.perform(self, player)
